src/core/lib/support/cpu_posix.cc

```diff
@@ -52,7 +53,27 @@ unsigned gpr_cpu_current_cpu(void) {
      most code that's using this is using it to shard across work queues though,
      so here we use thread identity instead to achieve a similar though not
      identical effect */
-  return (unsigned)GPR_HASH_POINTER(&magic_thread_local, gpr_cpu_num_cores());
+  static auto DeleteValue = [](void *value_ptr) {
+    unsigned int *value = static_cast<unsigned int *>(value_ptr);
+    if (value) {
+      delete value;
+    }
+  };
+  static pthread_key_t thread_id_key;
+  static int thread_id_key_create_result __attribute__((unused)) =
+      pthread_key_create(&thread_id_key, DeleteValue);
+  // pthread_t isn't portably defined to map to an integral type. So keep track
+  // of thread identity explicitly so hashing works reliably.
+  static std::atomic<unsigned int> thread_counter(0);
+
+  unsigned int *thread_id =
+      static_cast<unsigned int *>(pthread_getspecific(thread_id_key));
+  if (thread_id == nullptr) {
+    thread_id = new unsigned int(thread_counter++);
```

Comment:
Sorry, we can't use new and delete into the core: https://github.com/grpc/grpc/blob/master/doc/core/moving-to-c%2B%2B.md

Please change to using gpr_malloc and gpr_free instead.